redash/models.py

```diff
@@ -1,238 +1,140 @@
+            self._object = session.query(object_class).filter(
+                object_class.id == self.object_id).first()
+            return self._object
 
-db = Database()
+    @object.setter
+    def object(self, value):
+        self._object = value
+        self.object_type = value.__class__.__tablename__
+        self.object_id = value.id
 
 
-# Support for cast operation on database fields
-@peewee.Node.extend()
-def cast(self, as_type):
-    return peewee.Expression(self, '::', peewee.SQL(as_type))
+# # Support for cast operation on database fields
+# @peewee.Node.extend()
+# def cast(self, as_type):
+#     return peewee.Expression(self, '::', peewee.SQL(as_type))
 
 
-class JSONField(peewee.TextField):
-    def db_value(self, value):
+class PseudoJSON(TypeDecorator):
+    impl = db.Text
+    def process_bind_param(self, value, dialect):
         return json_dumps(value)
-
-    def python_value(self, value):
+    def process_result_value(self, value, dialect):
         if not value:
             return value
         return json.loads(value)
 
 
-class BaseModel(MeteredModel):
-    class Meta:
-        database = db.database
-
-    @classmethod
-    def get_by_id(cls, model_id):
-        return cls.get(cls.id == model_id)
-
-    def pre_save(self, created):
-        pass
-
-    def post_save(self, created):
-        # Handler for post_save operations. Overriding if needed.
-        pass
-
-    def save(self, *args, **kwargs):
-        pk_value = self._get_pk_value()
-        created = kwargs.get('force_insert', False) or not bool(pk_value)
-        self.pre_save(created)
-        super(BaseModel, self).save(*args, **kwargs)
-        self.post_save(created)
-
-    def update_instance(self, **kwargs):
-        for k, v in kwargs.items():
-            # setattr(model_instance, field_name, field_obj.python_value(value))
-            setattr(self, k, v)
-
-        # We have to run pre-save before calculating dirty_fields. We end up running it twice,
-        # but pre_save calls should be very quick so it's not big of an issue.
-        # An alternative can be to recalculate dirty_fields, but it felt more error prone.
-        self.pre_save(False)
-
-        self.save(only=self.dirty_fields)
-
-
-class ModelTimestampsMixin(BaseModel):
-    updated_at = DateTimeTZField(default=datetime.datetime.now)
-    created_at = DateTimeTZField(default=datetime.datetime.now)
-
-    def pre_save(self, created):
-        super(ModelTimestampsMixin, self).pre_save(created)
-        self.updated_at = datetime.datetime.now()
-
-
-def _simple_value(v):
-    if isinstance(v, BaseModel):
-        return v.id
-
-    return v
+class TimestampMixin(object):
+    updated_at = Column(db.DateTime(True), default=db.func.now(),
+                           onupdate=db.func.now(), nullable=False)
+    created_at = Column(db.DateTime(True), default=db.func.now(),
+                           nullable=False)
 
 
 class ChangeTrackingMixin(object):
```

Comment:
I found out why the event didn't trigger and implemented it: https://github.com/getredash/redash/commit/e8739b39678db57a1ab31ef4d2747294d448590c

If you have no comments, I will push this change to your branch.

The part I'm not happy about is how we deduce the user who changed the object, but I'm not sure it's that bad. Eventually this code is only relevant for the API, which is Flask based... and I added some safeguards to make sure it doesn't cause harm outside of Flask context.